### PR TITLE
validator server as k8s external admission hook

### PR DESCRIPTION
**Pull request by jmuk**
This PR adds the validation server as an external admission webhook
of kubernetes. This also introduce the structural validation
through protobuf definition.

The integration with referential validation is not yet done, that's
a TODO.

Also this adds a new subcommand to mixs. mixs validator will start
a webserver for the webhook.

**Release note**:

```release-note
validator server is added as a subcommand of mixs. Only structural validations right now.
```

---
This change is [Reviewable](https://reviewable.io/reviews/istio/mixer/1218)

### CODE REVIEW

**Review by ayj — COMMENTED** [review]
What was your plan for enabling this validation service via k8s webhook with GKE workaround? Use istioctl and non-HTTPS variant until k8s 1.8 with option of both schemes coexist for while?

**Comment by ayj** [comment]
I wrote a script to automate the GKE workaround + cert generation for Pilot admission webhook (see [here](https://github.com/istio/pilot/pull/1158/files#diff-8538e1bc18a904c05a69a1d9122497f2)). We could add something similar for Mixer validation if that makes sense. This may not really be suitable for e2e presubmit tests, but does provide a mechanism for users to experiment with the admission webhook validation before k8s 1.8 is available on GKE.

File: pkg/config/crd/validatorserver.go [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+	glog.V(7).Infof("received: %s", body)
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		glog.Errorf("contentType=%s, expect application/json", contentType)
```
[the remark this answers is not in the document]
Comment (jmuk) [reply pkg/config/crd/validatorserver.go]: done

File: pkg/config/crd/validatorserver_test.go [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+func sendRecv(h http.Handler, ctype string, in []byte) []byte {
+	inBuf := bytes.NewBuffer(in)
+	outBuf := &mockWriter{bytes.NewBuffer(nil)}
+	req, err := http.NewRequest("GET", "/", inBuf)
+	req.Header.Add("Content-Type", "application/json")
+	if err != nil {
+		return nil
+	}
+	h.ServeHTTP(outBuf, req)
```
[the remark this answers is not in the document]
Comment (jmuk) [reply pkg/config/crd/validatorserver_test.go]: done, thanks!

**Comment by jmuk** [comment]
@ayj, your script should be really helpful and work well with my patch too, thanks for the headsup. Will imitate it (or maybe it's worth having it in istio/istio repository?)

**Comment by jmuk** [comment]
reviewers, please take a look; I've confirmed this is working with @ayj's workaround script.

**Comment by jmuk** [comment]
@mandarjog ping

**Comment by jmuk** [comment]
@mandarjog ping

**Comment by geeknoid** [comment]
Do we still want this PR?

**Comment by jmuk** [comment]
Yes, absolutely, this is a part of validation for mixer config. This is still needed and I'm still waiting for reviews.

**Comment by geeknoid** [comment]
This repo is no longer accepting PRs. Please resubmit this change to the istio/istio repo.

Thanks.

**Pull request closed without merging**